ENH: Convexhull visiblefacets

rtshort opened the pull request:
This pull request has three unrelated changes to qhull interface.

1.  The important change.  Qhull has the ability to create a list of facets in the convex hull that are visible from a point via the QGn and QG-n flags.  The current qhull.pyx allows the user to use the options but doesn't provide enough information to actually extract the visible info.  The qhull library returns a list of facets and each facet has a "good" flag.  All this modification does is return the "good" flag for each facet in an array called good.

I am working up some unit tests, but there are some test programs in https://github.com/rtshort/spatial.git.

2.  In both the Delaunay and Voronoi classes I added a flag to the class that indicates whether this was a furthest site diagram or not.  I did this because carrying the input flag furthest_site along into the downstream processing (e.g. plotting or pole of inaccessibility computations) was error prone and clunky.  This is not terribly important, just a convenience.

3.  There seems to be an error in the docstrings for Voronoi - the docs said 1 finite and 4 finite ridges.  I believe I corrected this.

tylerjereddy commented on scipy/spatial/qhull.pyx:
```diff
@@ -541,6 +542,7 @@ cdef class _Qhull:
         cdef int i, j, ipoint, ipoint2, ncoplanar
         cdef object tmp
         cdef np.ndarray[np.npy_int, ndim=2] facets
+        cdef np.ndarray[np.npy_int, ndim=1] good
```
Probably ok for now -- we're trying to encourage new Cython to use [typed memoryviews](https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html), but the rest of the code here is using the original approach, so that's maybe fine for now.

rtshort replied:
I am always interested in learning, but I don't know much about Cython.

rtshort commented on scipy/spatial/qhull.pyx:
```diff
@@ -2470,7 +2487,7 @@ class Voronoi(_QhullUser):
            [ 1.5,  1.5]])
 
-    There is a single finite Voronoi region, and four finite Voronoi
+    There is a single finite Voronoi region, and eight infinite Voronoi
```
Let me reply to the supposed docstring error in the morning.  I could very well be wrong.  Actually, I don't think so but there is no reason to be in such a hurry that it gets worse rather than better.

The "good" thing seems to be really solid.  I have been using it to play around with the inner/outer hull thing we talked about on a different thread (Spherical Voronoi diagrams with all the points in a single hemisphere) and it seems to be rock solid.  Since I am just using qhull this isn't a bit surprise.  I have not figured out how to do self tests in this framework, but I do have several tests that, when I figure it out, I will add to the pull request.

If you clone the repository I quoted above ( https://github.com/rtshort/spatial.git) I have a really cute 2d test program and a much less sophisticated 3d test program.

rtshort commented on scipy/spatial/qhull.pyx:
```diff
@@ -2517,6 +2534,8 @@ class Voronoi(_QhullUser):
                        incremental=incremental)
         _QhullUser.__init__(self, qhull, incremental=incremental)
 
+        self.furthest_site = furthest_site
```
Ok. I can break this down into three separate pull requests.

rtshort replied:
I don't know how to withdraw a PR and start new ones though.

tylerjereddy replied:
What you did is ok for now, we can clean up the git history later & best to leave this PR open.

rtshort commented:
I reverted the changes for the two minor updates.  I don't seem to be able to edit the pull request, but for now only the major issue is in the fork.  I will do the other two one at a time when we finish this.

rtshort commented on scipy/spatial/qhull.pyx:
```diff
@@ -2288,6 +2295,9 @@ class ConvexHull(_QhullUser):
         triangulation due to numerical precision issues.
+    good : Boolean array indicated which facets (simplices) are good.
+        Used with options that compute good facets, e.g. QGn and 
+        QG-n.
```
As for the not-visible from inside the hull, the qhull documentation doesn't seem to describe this, but my interpretation is that the OUTER edge has to be visible and the inner edge doesn't count.  It also jives exactly with what I want to happen.  :->.  I was a really surprised by this at first but it makes sense if you take my interpretation.

Also, I have a lot of interesting examples to include in the documentation.  I will post a couple here.  3d examples are harder, but I have a couple where all the points are on the unit sphere that are kind of cool.  My programs are all dynamic in that I use the mouse or keyboard to cycle through facets and it will take a little work to make static documentation.  I only get to work on this here and there so it will take a couple of days but I will do it.

This really is a cool feature.  I am glad you are exercising it beyond what I thought of doing.

tylerjereddy replied:
I suspect we'd just want one simple / concise example for the docstring. If there are a lot of more interesting cases, I suppose those might be placed in a tutorial. The docs may need to clarify the inner / outer face distinction too, as this may cause confusion otherwise.

I provided a likely suitable file for placing the unit tests above.

rtshort replied:
1.  I see the unit test reference now.  Sorry for the noise.
2. You are right about simple docstrings of course.
3. My comment about incorrect docstrings (issue 3 in the pull request) was incorrect and you are correct.  Just didn't read it right.

I don't know how much time I will have before the weekend but I can rebase, make some of your suggested changes, etc. as soon as I can.

rtshort commented:
I made a passel of changes, hopefully everything we discussed.  A list follows.  I am not especially git/github literate so I hope I didn't dork things up horribly.

typed memoryviews: You didn't ask for a change, but doesn't make sense to have two different approaches and I am no way no how competent to make these changes.

docstrings: Moved "good" stuff below the coplanar stuff.
                  Added "good" example to ConvexHull - just stole yours and made it a little shorter.
                 Enhanced the "good" documentation and changed "indicated" to "indicating" (good catch).
                 I have a little trepidation here since I am providing my interpretation of the very sparse 
                 qhull documentation, but overall I think this is an improvement.

Added "versionadded".  Since I don't know what version (if ever) this stuff will be added this is incomplete.

I took the good attribute out of Delaunay.  All facets are always not-good all the time.

I added unit tests.

Not done: I have not played with incremental stuff.  It should work but obviously should be tested.  I will do that over the next few days.

rtshort commented:
Note: Incremental processing only works if "restart=True".  This will take a little thought.

tylerjereddy commented on scipy/spatial/qhull.pyx:
```diff
@@ -2335,6 +2356,56 @@ class ConvexHull(_QhullUser):
+    >>> convex_hull_plot_2d(hull, ax=ax)
+    >>> plt.show()
+
+    Note that, in the above example, if the point have been
```
```suggestion
    Note that, in the above example, if the point had been
```

tylerjereddy commented on scipy/spatial/qhull.pyx:
```diff
@@ -2433,6 +2505,10 @@ class Voronoi(_QhullUser):
+    furthest_site
+        True if this was a furthest site tesselation and False if not.
+
+        .. versionadded:: 1.3.x
```
It looks like the git history / changes got mangled a bit -- this stuff has creeped back in now.

tylerjereddy commented on scipy/spatial/tests/test_qhull.py:
```diff
@@ -701,6 +701,43 @@ def test_volume_area(self):
+                           [0.4, 0.4],
+                           [0.4, 0.2],
+                           [0.3, 0.6]])
+        hull = ConvexHull(points=points, qhull_options='QG4')
```
if you check the CI results and test examples above, you'll see that `qhull.ConvexHull` is needed here and in the other cases you've added

rtshort replied:
I have no idea how to read that.

tylerjereddy commented:
> Note: Incremental processing only works if "restart=True". This will take a little thought.

Are you saying that the visibility feature you're adding here only works with `incremental` when `restart=True`? Does it raise an error or segfault or silently return something incorrect otherwise?

rtshort commented:
The facets are updated but apparently the "good" array needs some work to update it as well.  I probably won't try to fix it.  I have no need for this feature and it is going to take a lot of digging to figure it out.

For the next six or eight weeks the time I will have available to work on this is going to be very slim.  I expect I will just have to drop this whole effort until the other project is finished.

tylerjereddy commented:
We could think about adding the feature only for the non-incremental case initially & open an issue with the idea to support the incremental addition of points in combination with visibility later on. Divide and conquer, maybe!

If you are too swamped I could also try to chip away at the PR a bit by pushing a few commits here and there & preserving your git history for now. I don't think there's a big rush on the feature in any case.

rtshort commented:
Sorry for the slow response.  Travel.  I will be doing a lot of that for the next couple of months.

I think your notion of how to do the incremental thing is a good one.  Document it and leave it for later.

I agree that this feature is not a rush.  It is very possible that I am the only person who ever wanted it!  I will have little free time until May or so but that isn't the same as none.  You are more than welcome to poke and prod at the repo.

tylerjereddy commented:
I rebased & pushed a commit that should allow the new `ConvexHull` visibility unit tests to pass & that restores the PR to a state where only the changes focused on the major new feature (`ConvexHull` facet visibility) should be present in the diff.

Still on the to-do:
- clean up the docstring additions a bit
- decide how to proceed re: incremental visibility

tylerjereddy commented:
Ok, I've added support for incremental visibility updates & substantially expanded the unit tests for this and other cases.

I tried to clean up the docstring a bit too.

This is ready for broader review I think.

tylerjereddy commented:
Looks like there's one issue specific to 32-bit linux.

tylerjereddy commented on scipy/spatial/qhull.pyx:
```diff
@@ -2290,6 +2300,19 @@ class ConvexHull(_QhullUser):
         triangulation due to numerical precision issues.
 
         If option "Qc" is not specified, this list is not computed.
+    good : A one-dimensional Boolean array indicating which facets are
```
This line should be adjusted to more closely match our usual standard. See the issue in [the circle ci render](https://11925-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.spatial.ConvexHull.html#scipy.spatial.ConvexHull) at time of writing.

Beyond that, the `ConvexHull` visibility additions look pretty nice there, but I'm biased.

tylerjereddy replied:
I've tried adjusting this -- we'll see if it looks slightly better once the CI docs are built

tylerjereddy commented on scipy/spatial/tests/test_qhull.py:
```diff
@@ -701,6 +701,120 @@ def test_volume_area(self):
+         np.array([False, False, True, False, False], dtype=bool)),
+        # degeneracy will also count as invisibility with current
+        # Qhull config
+        (np.array([[0.3, 0.6]]),
```
This is the problem case -- adding the degenerate generator incrementally fails only on 32-bit Linux.

tylerjereddy replied:
My current attempted fix here (which seems to work in a 32-bit Docker container) is to add `1e-16` to the y-value such that the query point is just barely inside the Convex Hull formed by adding this new point, near the limit of double precision.

It seems, albeit speculatively, that *exact* degeneracy may be handled slightly differently on 32-bit vs. 64-bit POSIX platforms, even with `m32` compile flag addition. It could be that our default usage of `Qt` Qhull option for all scenarios is the culprit on that, but that's generally a desirable option & I don't think we should add a special hook to disable graceful degeneracy handling for this obscure scenario ("is something visible if you're exactly on top of it?").

So my suggestion is that as far as visibility is concerned, if you add a generator that is *exactly* degenerate with your query point, it is fair game for Qhull to effectively move that point a bit and return **visibility** results that may vary on different architectures.

Maybe a concise doc addition is warranted, but on the other hand we already document that `Qt` is **always** enabled.

rgommers commented:
> This is ready for broader review I think.

Overall this looks like a good addition. [html docs](https://11925-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.spatial.ConvexHull.html#scipy.spatial.ConvexHull) also look fine. Were you looking for specific input? If both of you are happy and there are no comments for some days, I'd say go ahead and merge (after test failure is fixed)

rtshort commented:
Sorry I have been off the grid for a bit.  I have some time this weekend and will look at what you have done but it looks great to me at a glance.  One question though: how do I get the html from the code?

rgommers commented:
> how do I get the html from the code?

if you want to build the html docs for yourself: `cd doc && make html`.

tylerjereddy commented:
I'll aim to merge in about two weeks (March 23) if there are no unresolved issues brought up before then. 

I tried a little [property-based testing](https://github.com/tylerjereddy/scipy-property/blob/master/tests/test_qhull.py#L15) to probe for major weaknesses and didn't find any on a first pass. Of course, it is easy to cause `QhullError` more broadly when generators aren't in general position (coplanarity, degeneracy, etc.).

tylerjereddy commented:
The pypy3 failure looks like a connection / 404 issue that is completely unrelated; pypy3 CI testing is frequently problematic.

The rest of the CI is green after my fresh rebase this morning, I've given two weeks grace for additional review comments / issues, but no objections and looks ok to me after making a few adjustments, so in it goes. Thanks @rtshort 

I've updated the release notes at https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.3.0
